orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/support/TargetReferenceLinearStageSupport.groovy

```diff
@@ -36,33 +40,55 @@ abstract class TargetReferenceLinearStageSupport implements StageDefinitionBuild
   DetermineTargetReferenceStage determineTargetReferenceStage
 
   @Override
-  def List<Stage> aroundStages(Stage parentStage) {
-    return composeTargets(parentStage)
+  void beforeStages(@Nonnull Stage parent, @Nonnull StageGraphBuilder graph) {
+    log.warn("TargetReferenceLinearStageSupport is deprecated (execution: ${parent.execution.id})")
+
+    parent.resolveStrategyParams()
+
+    if (targetReferenceSupport.isDynamicallyBound(parent) && !parent.parentStageId) {
+      Map injectedContext = new HashMap(parent.context)
+      injectedContext.regions = new ArrayList(parent.context.regions)
+      graph.add {
+        it.type = determineTargetReferenceStage.type
+        it.name = "determineTargetReferences"
+        it.context = injectedContext
+      }
+    }
+  }
+
+  @Override
+  void afterStages(@Nonnull Stage parent, @Nonnull StageGraphBuilder graph) {
+    parent.resolveStrategyParams()
+
+    List<Stage> stages = composeTargets(parent, graph)
+    for (int i = 1; i < stages.size(); i++) {
+      graph.connect(stages.get(i - 1), stages.get(i))
+    }
   }
 
-  List<Stage> composeTargets(Stage stage) {
-    stage.resolveStrategyParams()
+  List<Stage> composeTargets(Stage stage, StageGraphBuilder graph) {
     if (targetReferenceSupport.isDynamicallyBound(stage)) {
-      return composeDynamicTargets(stage)
+      return composeDynamicTargets(stage, graph)
     }
 
-    return composeStaticTargets(stage)
+    return composeStaticTargets(stage, graph)
   }
 
-  private List<Stage> composeStaticTargets(Stage stage) {
+  private List<Stage> composeStaticTargets(Stage stage, StageGraphBuilder graph) {
     def descriptionList = buildStaticTargetDescriptions(stage)
     if (descriptionList.empty) {
       throw new TargetReferenceNotFoundException("Could not find any server groups for specified target")
     }
     def first = descriptionList.remove(0)
     stage.context.putAll(first)
 
-    if (descriptionList.size()) {
-      return descriptionList.collect {
-        newStage(stage.execution, this.type, this.type, it, stage, SyntheticStageOwner.STAGE_AFTER)
+    return descriptionList.collect { description ->
```

Comment:
do we actually still need to return anything here or can the method be `void`?

Reply:
We do, for the reason of being able to connect up the stages correctly, unless we don't really care about connects (was just keeping with backwards compat).

Reply:
Shouldn't it connect them inside of this method?

Reply:
Mm... yeah. Probably should. Will change.